### DISCUSSION TITLE
Access externalMedia etc from Clip

### DISCUSSION THREAD

**Discussion by alatdneg** [AcademySoftwareFoundation/OpenTimelineIO#1616] [Q&A] [answered]
Hi, I'm looking at adding support for opentimelineio in xStudio, using the C++ interface.

I'm trying to access the media paths on the Clip objects via the media_references() function, this return's a list of mediaReferences, but I'm unsure how I determine / access the real objects.

I'm guessing the underlying classes are MissingReference ExternalReference etc. So I need to workout what it actual is and then dynamic_cast it ?

There doesn't seem to be any examples in the code showing this.

Any help would be most appreciated ! 

cheers

**Comment by darbyjohnston** [comment, marked as answer]
Yes, you can dynamically cast the media references something like this:
```
std::string mediaURL;
if (auto externalRef = dynamic_cast<const otio::ExternalReference*>(clip->media_reference()))
{
    mediaURL = externalRef->target_url();
}
else if (auto imageSequenceRef = dynamic_cast<const otio::ImageSequenceReference*>(clip->media_reference()))
{
    std::stringstream ss;
    ss << imageSequenceRef->target_url_base() <<
        imageSequenceRef->name_prefix() <<
        std::setfill('0') << std::setw(imageSequenceRef->frame_zero_padding()) <<
        imageSequenceRef->start_frame() <<
        imageSequenceRef->name_suffix();
    mediaURL = ss.str();
}
```

Here I am using ```clip->media_reference()``` which will return the main, or "active", media reference. The list of media references is for supporting multiple media files per clip, for example multiple proxy resolutions. Very useful but still a relatively new feature that may not be widely used yet.

**Reply by alatdneg** [reply]
Cheers!
